### PR TITLE
Revision of the MoCo SSL model

### PR DESCRIPTION
**Pull request by senarvi**
## What does this PR do?

Related to issue #839 that discusses a major revision of Bolts, this pull request refactors the Momentum Contrast (MoCo) SSL model. That issue will probably become impossible to follow if all the details are discussed there, so I'll write my suggestions below. I would be happy to hear some feedback.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by senarvi** [comment]
@Atharva-Phatak SimCLR is a very similar method, so it makes sense to coordinate our efforts, and maybe other SSL methods too, so that we could make the interfaces as consistent as possible. We could also place both models in `pl_bolts.models.ssl.contrastive` module in order to be able to share code, if that's a good idea.

I think it would be important to make the models generic so that one can train
* any network, instead of selecting one of a set of options
* using any projection head
* using any optimizer and learning rate schedule
* using any data loader and transforms

Issues #904 already suggests moving transforms to a central location. I don't know if we need separate transforms for each model. More or less identical transforms are used with both methods, and the user doesn't necessarily want to stick to the default transforms anyway.

We should fix the format that the data loader should produce for contrastive learning. The format could be as similar as possible to the format that we use for classification and object detection. At least for object detection we use a tuple `(images, targets)`, where `images` is a list of image tensors and SSL models can simply ignore `targets`. In this pull request `images` is expected to be a nested list `[[image1A, image1B], [image2A, image2B], [image3A, image3B], ...]`.

I added an example CLI application using LightningCLI, but IMO there's no point in trying to create a comprehensive training tool in each module.

I renamed the class to MoCo. I think v2 just adds a projection head, so the same class supports both versions.

**Comment by senarvi** [comment]
The old way of detecting whether DDP is in use didn't work, since the DDP strategy class is now `DDPStrategy`, not `DDPPlugin`. Using `isinstance()` is pretty error-prone. I wonder if we could use `if torch.distributed.is_available() and torch.distributed.is_initialized()`, or is it important to detect DDP as opposed to other distributed training strategies?

**Comment by Atharva-Phatak** [comment]
@senarvi Perfect, that is what my intention is essentially we add Projection heads and Losses required by SSL models. Backbones can be implemented by the user if they are custom or else imported from timm or something similar. 

All in all here is what I suggest

1) Single place for SSL projection heads
2) Single place for SSL losses
3) Single place for transforms as well.
3) General PL module for training any custom SSL model :)

**Comment by otaj** [comment]
Hi, @senarvi, @Atharva-Phatak. I fully support @Atharva-Phatak's suggestion, if you can coordinate on that, that would be great!

Btw, @senarvi, since your fork is on an organization account (`groke-technologies`) and not on personal account, noone except people belonging to that organization have write access to it. While that makes sense, I'd say it makes it hard for us maintainers to help you with those PRs since we cannot merge master.

**Comment by senarvi** [comment]
@otaj ok that explains it. But it's fine, I'll just keep merging master myself.

**Comment by ArnolFokam** [comment]
> @senarvi Perfect, that is what my intention is essentially we add Projection heads and Losses required by SSL models. Backbones can be implemented by the user if they are custom or else imported from timm or something similar.
> 
> All in all here is what I suggest
> 
> 1. Single place for SSL projection heads
> 2. Single place for SSL losses
> 3. Single place for transforms as well.
> 4. General PL module for training any custom SSL model :)

Happy to help with this as well. Something we might want to be cautious of (if we follow this route) is the compatibility between the transforms and losses of various SSL methods.

**Comment by Borda** [comment]
@senarvi seems I can't push changes to this branch/PR, could you pls merge master here? 🐿️

**Comment by Borda** [comment]
@senarvi how is it going here? :rabbit:

**Comment by Borda** [comment]
FYI: transforms were moved to on sub-package

**Comment by senarvi** [comment]
@Borda thanks for reminding me. I noticed that the transforms were moved to a different location, but there seems to be something else. Maybe a circular import. I'm looking into it.

**Comment by Borda** [comment]
> Maybe a circular import. I'm looking into it.

Could be, mind allowing maintainers to push to your PR so can try to help? :)

**Comment by senarvi** [comment]
@Borda the problem is, this comes from an organization, so I don't know if it's possible to allow developers to push to the branch. Now I hope that I was able to merge it. Fingers crossed that the tests will pass.

**Comment by Borda** [comment]
One test failing, looks promising: FAILED `tests/models/self_supervised/test_models.py::test_moco - ValueError: Contrastive training expects a tuple or a list of transformed images, got Tensor.`

**Review by Borda — APPROVED** [review]

**Comment by senarvi** [comment]
> One test failing, looks promising: FAILED `tests/models/self_supervised/test_models.py::test_moco - ValueError: Contrastive training expects a tuple or a list of transformed images, got Tensor.`

Let me look into that tomorrow. It means that the test is not providing data in the correct format. Unfortunately I wasn't able to run the tests locally. I'll have to find out how to do that.

**Comment by senarvi** [comment]
@Borda the problem was that the data module used in the unit tests provided 4-dimensional tensors of image pairs, while the default ImageNet data module provided tuples of 3-dimensional tensors and that's what the input validator expected too. Now it supports both.

There was also a new unit test that checks the CLI scripts. It failed because ImageNet cannot be downloaded automatically, so I changed the CLI application to use CIFAR10 instead.

If I'm correct, now all the unit tests pass. I can only see some Docker errors now. But let me know if there's still something that needs fixing.

**Comment by Borda** [comment]
@senarvi the GPU start issue is fixed on master...

**Comment by senarvi** [comment]
@Borda merged master and now everything passes.

File: src/pl_bolts/models/detection/retinanet/retinanet_module.py [hunk 1]
```diff
@@ -136,13 +136,17 @@ def configure_optimizers(self):
 @under_review()
 def cli_main():
     try:  # Backward compatibility for Lightning CLI
-        from pytorch_lightning.cli import LightningCLI  # PL v1.9+
-    except ImportError:
-        from pytorch_lightning.utilities.cli import LightningCLI  # PL v1.8
+        import pytorch_lightning.cli
+
+        cli_class: Any = getattr(pytorch_lightning.cli, "LightningCLI")  # PL v1.9+
+    except Exception:
```
Comment (Borda) [inline src/pl_bolts/models/detection/retinanet/retinanet_module.py:142]: can we be more specific here with exception?

File: src/pl_bolts/models/detection/retinanet/retinanet_module.py [hunk 2]
```diff
@@ -136,13 +136,17 @@ def configure_optimizers(self):
...
+        import pytorch_lightning.utilities.cli
+
+        cli_class = getattr(pytorch_lightning.utilities.cli, "LightningCLI")  # PL v1.8
```
Comment (Borda) [inline src/pl_bolts/models/detection/retinanet/retinanet_module.py:145]:
```suggestion
        from pytorch_lightning.cli import LightningCLI  # PL v1.9+
    except ImportError:
        from pytorch_lightning.utilities.cli import LightningCLI  # PL v1.8
```

File: src/pl_bolts/models/detection/retinanet/retinanet_module.py [hunk 3]
```diff
@@ -136,13 +136,17 @@ def configure_optimizers(self):
...
 
     from pl_bolts.datamodules import VOCDetectionDataModule
 
-    LightningCLI(RetinaNet, VOCDetectionDataModule, seed_everything_default=42)
+    cli_class(RetinaNet, VOCDetectionDataModule, seed_everything_default=42)
```
Comment (Borda) [inline src/pl_bolts/models/detection/retinanet/retinanet_module.py:149]:
```suggestion
    LightningCLI(RetinaNet, VOCDetectionDataModule, seed_everything_default=42)
```

File: src/pl_bolts/models/detection/yolo/yolo_module.py [hunk 4]
```diff
@@ -614,4 +613,13 @@ def _resize(self, image: Tensor, target: TARGET) -> Tuple[Tensor, TARGET]:
 
 
 if __name__ == "__main__":
-    LightningCLI(CLIYOLO, ResizedVOCDetectionDataModule, seed_everything_default=42)
+    try:  # Backward compatibility for Lightning CLI
```
Comment (Borda) [inline src/pl_bolts/models/detection/yolo/yolo_module.py:616]: The same here...

**Comment by senarvi** [comment]
@Borda the thing that you commented on is a real PITA to get working so that mypy won't complain in any case. There's a lot of talk about this on GitHub, see for example [this](https://github.com/python/mypy/issues/8823). The problem is that mypy doesn't consider the `try..except` and complains about LightningCLI being already defined. If `# type: ignore` comments are added, it will complain about an unused 'type: ignore' comment, because always one of the comments is unnecessary. I'm not sure if there's a fix in the latest mypy. At least with mypy 1.1.1 the code that you suggested won't pass.

**Comment by senarvi** [comment]
@Borda also, the exception can be either an `ImportError` or an `AttributeError`, so it's an ugly workaround indeed.

**Comment by senarvi** [comment]
@Borda should we go like this? I'm fine with anything that makes mypy happy.

**Comment by Borda** [comment]
> I'm fine with anything that makes mypy happy.

lets mypy/typing address in separate PR :)

File: src/pl_bolts/models/detection/retinanet/retinanet_module.py (hunk 2 above)
Comment (Borda) [reply src/pl_bolts/models/detection/retinanet/retinanet_module.py:145]: @senarvi lets use standard import pattern

**Comment by senarvi** [comment]
@Borda fine. Actually it was necessary to only support the current location (pytorch_lightning.cli) to make the tests pass. Now it's clean.

**Review by Borda — APPROVED** [review]

**Pull request merged**